crates/ruff/src/jupyter/schema.rs

```diff
@@ -0,0 +1,256 @@
+//! The JSON schema of a Jupyter Notebook, entrypoint is [`JupyterNotebook`]
+//!
+//! Generated by <https://app.quicktype.io/> from
+//! <https://github.com/jupyter/nbformat/blob/16b53251aabf472ad9406ddb1f78b0421c014eeb/nbformat/v4/nbformat.v4.schema.json>
+//! Jupyter Notebook v4.5 JSON schema.
+//!
+//! Compared to the generated version, `Cell::id` is optional because it wasn't required <v4.5,
+//! I added `#[serde(deny_unknown_fields)]` where the schema had `"additionalProperties": false`
+//! and `#[serde(flatten)] pub other: BTreeMap<String, Value>` otherwise as initial part for
+//! round-trip support.
+
+use std::collections::{BTreeMap, HashMap};
+
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+
+/// The root of the JSON of a Jupyter Notebook
+///
+/// Generated by <https://app.quicktype.io/> from
+/// <https://github.com/jupyter/nbformat/blob/16b53251aabf472ad9406ddb1f78b0421c014eeb/nbformat/v4/nbformat.v4.schema.json>
+/// Jupyter Notebook v4.5 JSON schema.
+#[derive(Debug, Serialize, Deserialize)]
+pub struct JupyterNotebook {
+    /// Array of cells of the current notebook.
+    pub cells: Vec<Cell>,
+    /// Notebook root-level metadata.
+    pub metadata: JupyterNotebookMetadata,
+    /// Notebook format (major number). Incremented between backwards incompatible changes to the
+    /// notebook format.
+    pub nbformat: i64,
+    /// Notebook format (minor number). Incremented for backward compatible changes to the
+    /// notebook format.
+    pub nbformat_minor: i64,
+}
+
+/// Notebook raw nbconvert cell.
+///
+/// Notebook markdown cell.
+///
+/// Notebook code cell.
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(deny_unknown_fields)]
+pub struct Cell {
+    pub attachments: Option<HashMap<String, HashMap<String, SourceValue>>>,
```

Comment:
Is the difference between an empty `HashMap` (and an empty `Vec` for `outputs`) important or can we change the types to `HashMap<String<HashMap<String, SourceValue>>` and use `skip_if=HashMapp::is_empty`?

The same applies for other `Option<Collection>` usages in this file

Reply:
afaik it saying the field to be missing vs. the field is present but empty. it believe it makes a difference for round-trip serialization and json schema compliance